Store switching improvements

samiuelson opened the pull request:
Store switching improvements.

Closes: #7979
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
Problem:
1. The list of stores (websites) was populated initially from the DB and then fetched from API and re-rendered. During the re-render, a blink effect was visible even if the list remained the same.

2. If the user selected a new store before the list was refreshed (before the above re-render blink) their choice was reset after the refresh.

### Testing instructions
Run the `trunk` build and navigate to `Menu` > `Switch store`:
1.  Verify there is a blink visible on the list
2. Verify that the store choice is being reset if it's made before the blink

Then verify that the above issues are solved with this PR.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Fix updating selected site info

Selected site data was not updated properly. This commit added call to `SitePickerViewModel::selectedSite::set` function to update the selected site.

samiuelson committed:
Prevent site list from blinking on refresh

Fixed `DiffUtil.ItemCallback::areContentsTheSame` implementation.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerViewModel.kt:
```diff
@@ -364,6 +364,7 @@ class SitePickerViewModel @Inject constructor(
     }
 
     fun onSiteSelected(siteModel: SiteModel) {
+        selectedSite.set(siteModel)
```
Changing the `SelectedSite` value will trigger any observers of the `SelectedSite` class, and might even clear some caches (if I remember correctly), since we require the user to click on the "Continue" button on the screen to confirm the selection, we shouldn't set the `SelectedSite` until the user clicks on it.

Was this change done to fix the point 2? if yes, then I think the correct solution be to add a second `MutableLiveData` to store the `id` of the selected `SiteModel`, so it would be initiated using the `SelectedSite` object, then changed after the user's actions, then replace the usages of `SelectedSite` with it inside `displaySites`.

A better fix (just sharing a personal opinion 🙂, no need to spend any time on it) is to make the screen more reactive, the usage of the `dispalySites` is something that's not very intuitive, normally we should have two streams: `sites` (different than the one we have now, it would reference raw `SiteModel`), and `selectedSiteId`, then we should compose the state from these two streams, which means the `displaySites` logic would be invoked as a reaction to one of them changing, and not invoked directly.

samiuelson replied:
Thanks for the explanation. I'll work it out the way you suggested 👍🏼

hichamboushaba replied:
Thanks @samiuelson.
Just sharing an additional issue that's caused by this change that I forgot to mention, if the user opens the screen, then clicks on a new Site, then closes the screen, we will still change the selected store for them, which is not what we want 🙂.

hichamboushaba reviewed (commented):
Thanks Samuel for working on this, I left a comment to discuss the second change.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerAdapter.kt:
```diff
@@ -143,6 +143,14 @@ class SitePickerAdapter(
             newItem: SitesListItem
-        ): Boolean = oldItem == newItem
+        ): Boolean = when {
+            oldItem is WooSiteUiModel && newItem is WooSiteUiModel -> {
+                oldItem.site.getSiteName() == newItem.site.getSiteName()
+            }
+            oldItem is NonWooSiteUiModel && newItem is NonWooSiteUiModel -> {
+                oldItem.site.getSiteName() == newItem.site.getSiteName()
+            }
+            else -> oldItem == newItem
```
Regarding the blinking here, while this fixes the issue, it has some issues because we are basically ignoring any content changes between the objects besides the site name, for example now changing the selected site in the screen doesn't have any effect (UI wise).

The real fix would be to implement a good `equals` implementation for `SiteModel`, or just make it a `data class` if it was in Kotlin 😄.
An easier solution for us is to update the `equals` implementation of `WooSiteUiModel` and `NonWooSiteUiModel` to check for what we need in the UI:
1. The site name.
2. The site URL.
3. And the status of `isSelected` inside `WooSiteUiModel`.

With these implemented, `oldItem == newItem` should be enough without any blinking I think.

WDYT about this idea @samiuelson?

samiuelson replied:
Yes. That's the right way to go. Good catch, thanks for pointing this out @hichamboushaba!

samiuelson committed:
Improve site list items comparing

Implemented `equals` and `hashCode` methods. Reverted `areContentsTheSame` implementation to compare objects (not specific props).

samiuelson committed:
Fix store switching logic

Storing selected site id in a separate variable

samiuelson commented:
@hichamboushaba thank you for your in-depth review! I've fixed the issues you pointed out and the PR's ready for another round of review.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerViewModel.kt:
```diff
@@ -666,12 +671,25 @@ class SitePickerViewModel @Inject constructor(
+                other.isSelected == isSelected &&
+                other.site.url == site.url
+
+            override fun hashCode(): Int = Objects.hash(site.siteId, site.selfHostedSiteId, site.url, isSelected)
```
Just nitpicking here, this implementation doesn't respect the second point of `hashCode` contract:

> - If two objects are equal according to the equals() method, then calling the hashCode method on each of the two objects must produce the same integer result.

And this is because it includes checks on `siteId` while the `equals` function doesn't, meaning that if the `siteId` gets updated, even though `equals` would still be `true`, the `hashCode` would return different results.
WDYT about updating `equals` to check `siteId` as well?

And regarding the `selfHostedSiteId`, this is not very clear from its name, but it's a field specific to sites using XMLRPC in the app, which we don't need ourselves, for us it's always `0`, so we can probably remove it from the check.

samiuelson replied:
Good catch, thanks. I fixed that in both `WooSiteUiModel` and `NonWooSiteUiModel`.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerViewModel.kt:
```diff
@@ -78,6 +79,8 @@ class SitePickerViewModel @Inject constructor(
     private val _sites = MutableLiveData<List<SitesListItem>>()
     val sites: LiveData<List<SitesListItem>> = _sites
 
+    private val selectedSiteId: MutableLiveData<Int> = MutableLiveData()
```
Another np, and sorry for not mentioning it on the initial suggestion, feel free to ignore it though, it's not a blocking point.

With this logic, the site selection gets lost after process-death, we can solve this by replacing this with a `LiveData` that's backed up with `SavedStateHandle`:
```kotlin
private val selectedSiteId: MutableLiveData<Int> = savedState.getLiveData("selected-site-id")
```

If we do so, we'll just need to adjust the initialization logic in the `init` block to avoid overwriting the saved value, something like this would work I think:

```kotlin
if (selectedSiteId.value == null && selectedSite.exists()) {
    selectedSiteId.value = selectedSite.getSelectedSiteId()
}
```

samiuelson replied:
Good idea! Thanks

hichamboushaba reviewed (approved):
Nice work @samiuelson, I tested it and it works well 👍.

I left two np comments (sorry 😓), feel free to decide whether they are worth addressing or not, they are not blockers for the PR.

samiuelson committed:
Fix `hashcode` and `equals` implementations

In `WooSiteUiModel` and `NonWooSiteUiModel` to respect hashCode's contract.

samiuelson committed:
Back up `selectedSiteId` with `SavedStateHandle`
